rasa/shared/core/trackers.py

```diff
@@ -461,46 +458,11 @@ def applied_events(self) -> List[Event]:
                 self._undo_till_previous_loop_execution(
                     event.action_name, applied_events
                 )
-            elif isinstance(event, UserUttered):
-                # update event's featurization based on the future event
-                use_text_for_featurization = self._define_user_featurization(
-                    events_as_list[i + 1 :]
-                )
-                if event.use_text_for_featurization is None:
-                    event.use_text_for_featurization = use_text_for_featurization
-                # update event's entities based on the future event
-                entities = self._define_user_entities(events_as_list[i + 1 :])
-                if entities is not None:
-                    for entity in entities:
-                        if entity not in event.entities:
-                            event.entities.append(entity)
-
-                applied_events.append(event)
-            else:
+            elif not isinstance(event, DefinePrevUserUttered):
```

Comment:
@wochinge for some reason this change breaks everything. If I don't include `Define...` events in `applied_events` sometimes they're not applied. I'll remove it from here. Anyhow, this should be redesigned

Reply:
Mhm, then we should investigate why this breaks everything. That can only mean that events are not applied properly to the tracker.

Reply:
Maybe the problem is that your `apply_to` implementations manipulate `last_user_message` and `last_action` instead of the events `tracker.events` 🤔 They should refer to the same things, but maybe some stuff is actually deepcopying it.

Reply:
ah, now I understand it! Yes, you definitely have to still add the event as applied events.

Reply:
If you don't add it, then the events are not applied in the new tracker (as it doesn't get them) and everything will break

Reply:
but... even if I add them, there is still problems during evaluation. I patched it in featurizers. But, basically pandora has already opened. Let's carefully redesign it after release

Reply:
> then the events are not applied in the new tracker

but the new tracker should already get new updated `UserUttered`

Reply:
> but the new tracker should already get new updated UserUttered

There is no need to. The new tracker will call `apply_to` on each new event

Reply:
but then we do it million of times, doesn't feel efficient for me

Reply:
We actually need to do this for all of the events (e.g. `restarted` needs to reset a lot of state). I think the problem is rather in the featurization

Reply:
what do you mean? in core featurizers? Theoretically they should always receive `UserUttered` events where their featurization is decided

Reply:
I'm not sure about this. Anyway, let's close this discussion and have it in comment of the large PR